### ISSUE TITLE
UpstreamReadError on zip containing another zip

### ISSUE THREAD

**Issue by sireliah** [Majored/rs-async-zip#32]
Hi, I've been using the crate for copying directories over a zip stream and I noticed that the `read::stream::ZipFileReader` fails in case the compressed stream contains a file that is already compressed by deflate method. 

```rust
use async_zip::read::stream::ZipFileReader;
use tokio::fs::File;

#[tokio::main]
async fn main() {
    // Fails when file contains already compressed file
    // e.g. epub, odt, zip
    let mut file = File::open("with-zipfile.zip").await.unwrap();
    let mut zip = ZipFileReader::new(&mut file);
    while !zip.finished() {
        if let Some(reader) = zip.entry_reader().await.unwrap() {
            let entry = reader.entry();
            println!("{:?}", entry.name());
            let mut out = vec![];
            reader.copy_to_end_crc(&mut out, 1024).await.unwrap();
        }
    }
}
```

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: UpstreamReadError(Custom { kind: Other, error: DecompressError(General { msg: None }) })', src/main.rs:16:47
stack backtrace:
   0: rust_begin_unwind
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:517:5
   1: core::panicking::panic_fmt
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/panicking.rs:100:14
   2: core::result::unwrap_failed
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/result.rs:1616:5
   3: core::result::Result<T,E>::unwrap
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/result.rs:1298:23
   4: test_zip::main::{{closure}}
             at ./src/main.rs:16:13
   5: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/future/mod.rs:80:19
   6: tokio::park::thread::CachedParkThread::block_on::{{closure}}
             at /.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.0/src/park/thread.rs:267:54
   7: tokio::coop::with_budget::{{closure}}
             at /.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.0/src/coop.rs:102:9
   8: std::thread::local::LocalKey<T>::try_with
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/thread/local.rs:399:16
   9: std::thread::local::LocalKey<T>::with
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/thread/local.rs:375:9
  10: tokio::coop::with_budget
             at /.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.0/src/coop.rs:95:5
  11: tokio::coop::budget
             at /.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.0/src/coop.rs:72:5
  12: tokio::park::thread::CachedParkThread::block_on
             at /.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.0/src/park/thread.rs:267:31
  13: tokio::runtime::enter::Enter::block_on
             at /.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.0/src/runtime/enter.rs:152:13
  14: tokio::runtime::scheduler::multi_thread::MultiThread::block_on
             at /.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.0/src/runtime/scheduler/multi_thread/mod.rs:79:9
  15: tokio::runtime::Runtime::block_on
             at /.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.0/src/runtime/mod.rs:492:44
  16: test_zip::main
             at ./src/main.rs:10:5
  17: core::ops::function::FnOnce::call_once
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/ops/function.rs:227:5
```

And the file I'm trying to send:

```
$ zipinfo with-zipfile.zip 
Archive:  with-zipfile.zip
Zip file size: 89746 bytes, number of entries: 2
drwxr-xr-x  2.0 unx        0 bx stor 22-Sep-17 15:50 with-zipfile/
-rw-r--r--  2.0 unx    89326 bX defN 22-Sep-17 15:50 with-zipfile/olives.zip
2 files, 89326 bytes uncompressed, 89356 bytes compressed:  0.0%
```

I did some experiments and it seems to happen with many file formats that use deflate (specifically in the "normal" mode), for example: epub or odt.

Why would that matter that the stream contains a file that is also compressed? The `DecompressError` suggests that the error is somewhere in the `async-compression` flate2 call, but I couldn't narrow down where and why exactly. Some help would be appreciated!

**Comment by JonathanxD** [comment]
Hello, can you provide a reproducible sample file? I've tried zipping some text files and nesting the zip, as well as zipping epubs but sadly I couldn't reproduce this, they all worked well. 

In the past I've noticed that some zip files with specific sizes triggered some hidden bugs (because of pre-defined buffer sizes), so a sample file that doesn't work would help a lot, way better than trying arbitrary combinations (which didn't worked for me at least).

Also, I did noticed that your `olives.zip` has an `X` in the `zipinfo` column, which means it contains both an [extra field and the extended local header](https://man.archlinux.org/man/extra/unzip/zipinfo.1.en#DETAILED_DESCRIPTION).

This means that the file has extra information and its content wasn't known beforehand, so the *Data descriptor* was written after the compressed data as well. AFAIK, the most common scenario that a file has an extra field is for Zip64, which isn't supported yet (I did implemented the support, but it still not ready yet and I'm pretty busy those months), but I think that Zip64 is not supported by zip 2.0 since it was only documented in 4.5, so I'm kinda discarding the possibility that it is a zip64 file (even though the error is similar and some zip libraries doesn't follow the spec). 

The weird thing is that, if I produce a Zip file similar to yours, even though winrar and file-roller correctly handles it, Ark doesn't and `7z l` warns about wrong headers, but *async_zip* still correctly handles it, unlike your case, so I'm thinking that it may be feeding `flate2` with wrong bytes, the only way to know is having a failing zip I can test and debug.

**Comment by sireliah** [comment]
Thanks for looking into that. Here are two files that I was using for testing:
[with-zipfile.zip](https://github.com/Majored/rs-async-zip/files/9692338/with-zipfile.zip)
[with-epub.zip](https://github.com/Majored/rs-async-zip/files/9692339/with-epub.zip)

First is the one containing zip compressed jpg and the second one contains an ebook.

But there is something related to the fact how the zip files have been created. My OS is Linux Fedora and I created the zipfiles using Nautilus file browser. When I create the zipfile using `zip`, commandline:

```bash
$ zip -r with-zipfile2.zip with-zipfile/
```
the `async_zip` decompresses the contents without issues. Judging from that fact, it must be something with the extra flags you mentioned.

P.S. Previously I didn't include the deps, so here we go:

```toml
[package]
name = "test_zip"
version = "0.1.0"
edition = "2021"

[dependencies]
async_zip = "0.0.8"
tokio = {version = "1.20.1", features = ["full"] }
```

**Comment by JonathanxD** [comment]
Interesting, I've tested both zips and they fail on `0.0.8`, but it seems like it's already fixed by #28, try using the latest version from git:

```toml
[dependencies]
async_zip = { git = "https://github.com/Majored/rs-async-zip" }
```

And see if it helps.

Even though it's not the best advice to get the dependency directly from git/vcs, it may be a valid solution for the mid-term, at least until the next version is released. You can also pin a specific revision which guarantees it doesn't get a newer (and maybe incompatible) revision:

```toml
[dependencies]
async_zip = { git = "https://github.com/Majored/rs-async-zip", rev = "0f4bf7f09d2fef1236924a6af45a5641392faf6b" }
```

**Comment by sireliah** [comment]
If the problem was fixed on main, then great. I was planning to use git ref for the dependency, because of the other bug that was fixed here https://github.com/Majored/rs-async-zip/pull/31.

By the way, do you plan to release new crate version soon?

**Comment by Stefan99353** [comment]
A new crate version would be great. I am using the crate in a library I am making but I can't publish it using the git dependency. Version 0.0.8 is not working because of #28.

**Comment by Majored** [comment]
I've released `0.0.9` so I'll go ahead and close this. 👍
